clients/src/main/java/org/apache/kafka/clients/producer/RecordMetadata.java

```diff
@@ -81,8 +76,18 @@ public long timestamp() {
 
     /**
      * The checksum (CRC32) of the record.
+     *
+     * @deprecated As of Kafka 0.11.0. Because of the potential for message format conversion on the broker, the
+     *             computed checksum may not match what was stored on the broker, or what will be returned to the consumer.
+     *             It is therefore unsafe to depend on this checksum for end-to-end delivery guarantees. Additionally,
+     *             message format v2 does not support a record-level checksum. To maintain compatibility, a partial
+     *             checksum computed from the record timestamp, serialized key size, and serialized value size is
+     *             returned instead, but obviously this should not be depended on for end-to-end reliability.
      */
+    @Deprecated
     public long checksum() {
+        if (checksum == null)
+            this.checksum = DefaultRecordBatch.computePartialChecksum(timestamp, serializedKeySize, serializedValueSize);
```

Comment:
We should probably add a comment here and `ConsumerRecord` explaining that this will be `null` for message format >= 2 and hence why we call `DefaultRecordBatch`.